File: runtime/lua/test_hashtbl.lua
#!/usr/bin/env lua
-- Test Hashtbl module

-- Load dependencies in correct order
dofile("mlBytes.lua")  -- Provides caml_bit_and, caml_bit_or, caml_bit_lshift, caml_bit_rshift
dofile("ints.lua")     -- Provides caml_to_int32
dofile("compare.lua")  -- Provides caml_is_ocaml_string, caml_is_ocaml_block
dofile("hash.lua")     -- Provides caml_hash_* functions
dofile("hashtbl.lua")  -- Provides caml_hashtbl_* functions

local tests_passed = 0
local tests_failed = 0

local function test(name, fn)
  io.write("Test: " .. name .. " ... ")
  local success, err = pcall(fn)
  if success then
    tests_passed = tests_passed + 1
    print("✓")
  else
    tests_failed = tests_failed + 1
    print("✗")
    print("  Error: " .. tostring(err))
  end
end

local function assert_eq(actual, expected, msg)
  if actual ~= expected then
    error(msg or ("Expected " .. tostring(expected) .. ", got " .. tostring(actual)))
  end
end

local function assert_true(value, msg)
  if not value then
    error(msg or "Expected true, got false")
  end
end

local function assert_false(value, msg)
  if value then
    error(msg or "Expected false, got true")
  end
end

local function assert_error(fn, expected_msg)
  local success, err = pcall(fn)
  if success then
    error("Expected error but function succeeded")
  end
  if expected_msg and not string.find(tostring(err), expected_msg, 1, true) then
    error("Expected error containing '" .. expected_msg .. "', got: " .. tostring(err))
  end
end

print("====================================================================")
print("Hashtbl Module Tests")
print("====================================================================")
print()

print("Hash Table Creation Tests:")
print("--------------------------------------------------------------------")

test("create: empty hash table", function()
  local h = caml_hash_create()
  assert_eq(caml_hash_length(h), 0)
end)

test("create: with initial size", function()
  local h = caml_hash_create(32)
  assert_eq(caml_hash_length(h), 0)
  local stats = caml_hash_stats(h)
  assert_eq(stats.capacity, 32)
end)

test("create: with invalid size defaults to 16", function()
  local h = caml_hash_create(0)
  local stats = caml_hash_stats(h)
  assert_eq(stats.capacity, 16)
end)

print()
print("Add Tests:")
print("--------------------------------------------------------------------")

test("add: single binding", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "value1")
  assert_eq(caml_hash_length(h), 1)
end)

test("add: multiple bindings", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "value1")
  caml_hash_add(h, "key2", "value2")
  caml_hash_add(h, "key3", "value3")
  assert_eq(caml_hash_length(h), 3)
end)

test("add: duplicate keys (both bindings kept)", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "value1")
  caml_hash_add(h, "key1", "value2")
  assert_eq(caml_hash_length(h), 2)
end)

test("add: integer keys", function()
  local h = caml_hash_create()
  caml_hash_add(h, 1, "one")
  caml_hash_add(h, 2, "two")
  caml_hash_add(h, 3, "three")
  assert_eq(caml_hash_length(h), 3)
end)

test("add: mixed types", function()
  local h = caml_hash_create()
  caml_hash_add(h, 1, "number")
  caml_hash_add(h, "str", "string")
  caml_hash_add(h, 3.14, "float")
  assert_eq(caml_hash_length(h), 3)
end)

print()
print("Find Tests:")
print("--------------------------------------------------------------------")

test("find: existing key", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "value1")
  local value = caml_hash_find(h, "key1")
  assert_eq(value, "value1")
end)

test("find: non-existent key raises error", function()
  local h = caml_hash_create()
  assert_error(function()
    caml_hash_find(h, "missing")
  end, "Not_found")
end)

test("find: most recent binding with duplicate keys", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "old")
  caml_hash_add(h, "key1", "new")
  local value = caml_hash_find(h, "key1")
  assert_eq(value, "new")
end)

test("find: integer keys", function()
  local h = caml_hash_create()
  caml_hash_add(h, 42, "answer")
  local value = caml_hash_find(h, 42)
  assert_eq(value, "answer")
end)

print()
print("Find Opt Tests:")
print("--------------------------------------------------------------------")

test("find_opt: existing key", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "value1")
  local value = caml_hash_find_opt(h, "key1")
  assert_eq(value, "value1")
end)

test("find_opt: non-existent key returns nil", function()
  local h = caml_hash_create()
  local value = caml_hash_find_opt(h, "missing")
  assert_eq(value, nil)
end)

test("find_opt: most recent binding", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "old")
  caml_hash_add(h, "key1", "new")
  local value = caml_hash_find_opt(h, "key1")
  assert_eq(value, "new")
end)

print()
print("Remove Tests:")
print("--------------------------------------------------------------------")

test("remove: existing key", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "value1")
  caml_hash_remove(h, "key1")
  assert_eq(caml_hash_length(h), 0)
end)

test("remove: non-existent key (no effect)", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "value1")
  caml_hash_remove(h, "missing")
  assert_eq(caml_hash_length(h), 1)
end)

test("remove: one of duplicate keys", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "old")
  caml_hash_add(h, "key1", "new")
  caml_hash_remove(h, "key1")
  assert_eq(caml_hash_length(h), 1)
  -- Should find the remaining binding
  local value = caml_hash_find(h, "key1")
  assert_eq(value, "old")
end)

test("remove: after remove, key not found", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "value1")
  caml_hash_remove(h, "key1")
  assert_error(function()
    caml_hash_find(h, "key1")
  end, "Not_found")
end)

print()
print("Replace Tests:")
print("--------------------------------------------------------------------")

test("replace: new key", function()
  local h = caml_hash_create()
  caml_hash_replace(h, "key1", "value1")
  assert_eq(caml_hash_length(h), 1)
  local value = caml_hash_find(h, "key1")
  assert_eq(value, "value1")
end)

test("replace: existing key", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "old")
  caml_hash_replace(h, "key1", "new")
  assert_eq(caml_hash_length(h), 1)
  local value = caml_hash_find(h, "key1")
  assert_eq(value, "new")
end)

test("replace: removes all duplicate bindings", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "old1")
  caml_hash_add(h, "key1", "old2")
  caml_hash_add(h, "key1", "old3")
  caml_hash_replace(h, "key1", "new")
  assert_eq(caml_hash_length(h), 1)
  local value = caml_hash_find(h, "key1")
  assert_eq(value, "new")
end)

print()
print("Mem Tests:")
print("--------------------------------------------------------------------")

test("mem: existing key", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "value1")
  assert_true(caml_hash_mem(h, "key1"))
end)

test("mem: non-existent key", function()
  local h = caml_hash_create()
  assert_false(caml_hash_mem(h, "missing"))
end)

test("mem: after remove", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "value1")
  caml_hash_remove(h, "key1")
  assert_false(caml_hash_mem(h, "key1"))
end)

print()
print("Length Tests:")
print("--------------------------------------------------------------------")

test("length: empty hash table", function()
  local h = caml_hash_create()
  assert_eq(caml_hash_length(h), 0)
end)

test("length: tracks additions", function()
  local h = caml_hash_create()
  caml_hash_add(h, "k1", "v1")
  assert_eq(caml_hash_length(h), 1)
  caml_hash_add(h, "k2", "v2")
  assert_eq(caml_hash_length(h), 2)
  caml_hash_add(h, "k3", "v3")
  assert_eq(caml_hash_length(h), 3)
end)

test("length: tracks removals", function()
  local h = caml_hash_create()
  caml_hash_add(h, "k1", "v1")
  caml_hash_add(h, "k2", "v2")
  caml_hash_remove(h, "k1")
  assert_eq(caml_hash_length(h), 1)
end)

test("length: includes duplicate keys", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "v1")
  caml_hash_add(h, "key1", "v2")
  assert_eq(caml_hash_length(h), 2)
end)

print()
print("Clear Tests:")
print("--------------------------------------------------------------------")

test("clear: empties hash table", function()
  local h = caml_hash_create()
  caml_hash_add(h, "k1", "v1")
  caml_hash_add(h, "k2", "v2")
  caml_hash_clear(h)
  assert_eq(caml_hash_length(h), 0)
end)

test("clear: can reuse hash table", function()
  local h = caml_hash_create()
  caml_hash_add(h, "k1", "v1")
  caml_hash_clear(h)
  caml_hash_add(h, "k2", "v2")
  assert_eq(caml_hash_length(h), 1)
  local value = caml_hash_find(h, "k2")
  assert_eq(value, "v2")
end)

test("clear: on empty hash table", function()
  local h = caml_hash_create()
  caml_hash_clear(h)
  assert_eq(caml_hash_length(h), 0)
end)

print()
print("Iter Tests:")
print("--------------------------------------------------------------------")

test("iter: empty hash table", function()
  local h = caml_hash_create()
  local count = 0
  caml_hash_iter(h, function(k, v)
    count = count + 1
  end)
  assert_eq(count, 0)
end)

test("iter: single element", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "value1")
  local count = 0
  local found_key, found_value
  caml_hash_iter(h, function(k, v)
    count = count + 1
    found_key = k
    found_value = v
  end)
  assert_eq(count, 1)
  assert_eq(found_key, "key1")
  assert_eq(found_value, "value1")
end)

test("iter: multiple elements", function()
  local h = caml_hash_create()
  caml_hash_add(h, "k1", "v1")
  caml_hash_add(h, "k2", "v2")
  caml_hash_add(h, "k3", "v3")
  local count = 0
  caml_hash_iter(h, function(k, v)
    count = count + 1
  end)
  assert_eq(count, 3)
end)

test("iter: includes duplicate keys", function()
  local h = caml_hash_create()
  caml_hash_add(h, "key1", "v1")
  caml_hash_add(h, "key1", "v2")
  local count = 0
  caml_hash_iter(h, function(k, v)
    count = count + 1
  end)
  assert_eq(count, 2)
end)

print()
print("Fold Tests:")
print("--------------------------------------------------------------------")

test("fold: empty hash table", function()
  local h = caml_hash_create()
  local result = caml_hash_fold(h, function(k, v, acc)
    return acc + 1
  end, 0)
  assert_eq(result, 0)
end)

test("fold: count elements", function()
  local h = caml_hash_create()
  caml_hash_add(h, "k1", "v1")
  caml_hash_add(h, "k2", "v2")
  caml_hash_add(h, "k3", "v3")
  local count = caml_hash_fold(h, function(k, v, acc)
    return acc + 1
  end, 0)
  assert_eq(count, 3)
end)

test("fold: sum integer values", function()
  local h = caml_hash_create()
  caml_hash_add(h, "k1", 10)
  caml_hash_add(h, "k2", 20)
  caml_hash_add(h, "k3", 30)
  local sum = caml_hash_fold(h, function(k, v, acc)
    return acc + v
  end, 0)
  assert_eq(sum, 60)
end)

print()
print("Entries Iterator Tests:")
print("--------------------------------------------------------------------")

test("entries: empty hash table", function()
  local h = caml_hash_create()
  local count = 0
  for k, v in caml_hash_entries(h) do
    count = count + 1
  end
  assert_eq(count, 0)
end)

test("entries: iterate over elements", function()
  local h = caml_hash_create()
  caml_hash_add(h, "k1", "v1")
  caml_hash_add(h, "k2", "v2")
  caml_hash_add(h, "k3", "v3")
  local count = 0
  local keys = {}
  for k, v in caml_hash_entries(h) do
    count = count + 1
    keys[k] = v
  end
  assert_eq(count, 3)
  assert_eq(keys["k1"], "v1")
  assert_eq(keys["k2"], "v2")
  assert_eq(keys["k3"], "v3")
end)

print()
print("Keys/Values Tests:")
print("--------------------------------------------------------------------")

test("keys: empty hash table", function()
  local h = caml_hash_create()
  local keys = caml_hash_keys(h)
  assert_eq(#keys, 0)
end)

test("keys: returns all keys", function()
  local h = caml_hash_create()
  caml_hash_add(h, "k1", "v1")
  caml_hash_add(h, "k2", "v2")
  caml_hash_add(h, "k3", "v3")
  local keys = caml_hash_keys(h)
  assert_eq(#keys, 3)
end)

test("values: empty hash table", function()
  local h = caml_hash_create()
  local values = caml_hash_values(h)
  assert_eq(#values, 0)
end)

test("values: returns all values", function()
  local h = caml_hash_create()
  caml_hash_add(h, "k1", "v1")
  caml_hash_add(h, "k2", "v2")
  caml_hash_add(h, "k3", "v3")
  local values = caml_hash_values(h)
  assert_eq(#values, 3)
end)

print()
print("To Array Tests:")
print("--------------------------------------------------------------------")

test("to_array: empty hash table", function()
  local h = caml_hash_create()
  local arr = caml_hash_to_array(h)
  assert_eq(#arr, 0)
end)

test("to_array: with elements", function()
  local h = caml_hash_create()
  caml_hash_add(h, "k1", "v1")
  caml_hash_add(h, "k2", "v2")
  local arr = caml_hash_to_array(h)
  assert_eq(#arr, 2)
end)

print()
print("Resize Tests:")
print("--------------------------------------------------------------------")

test("resize: automatic on load factor", function()
  local h = caml_hash_create(4)
  -- Add enough elements to trigger resize (load factor > 0.75)
  for i = 1, 10 do
    caml_hash_add(h, "key" .. i, "value" .. i)
  end
  assert_eq(caml_hash_length(h), 10)
  -- All elements should still be findable
  for i = 1, 10 do
    local value = caml_hash_find(h, "key" .. i)
    assert_eq(value, "value" .. i)
  end
end)

test("resize: preserves all bindings", function()
  local h = caml_hash_create(2)
  for i = 1, 20 do
    caml_hash_add(h, i, i * 10)
  end
  assert_eq(caml_hash_length(h), 20)
  for i = 1, 20 do
    local value = caml_hash_find(h, i)
    assert_eq(value, i * 10)
  end
end)

print()
print("Complex Key Tests:")
print("--------------------------------------------------------------------")

test("complex: OCaml byte array keys", function()
  local h = caml_hash_create()
  local key1 = {72, 101, 108, 108, 111}  -- "Hello"
  local key2 = {87, 111, 114, 108, 100}  -- "World"
  caml_hash_add(h, key1, "greeting")
  caml_hash_add(h, key2, "earth")
  local v1 = caml_hash_find(h, key1)
  local v2 = caml_hash_find(h, key2)
  assert_eq(v1, "greeting")
  assert_eq(v2, "earth")
end)

test("complex: table keys with same structure", function()
  local h = caml_hash_create()
  local key1 = {1, 2, 3}
  local key2 = {1, 2, 3}  -- Different table, same structure
  caml_hash_add(h, key1, "first")
  local value = caml_hash_find(h, key2)
  assert_eq(value, "first")  -- Should find due to structural equality
end)

print()
print("Performance Tests:")
print("--------------------------------------------------------------------")

test("performance: many insertions", function()
  local h = caml_hash_create()
  for i = 1, 1000 do
    caml_hash_add(h, i, i * 2)
  end
  assert_eq(caml_hash_length(h), 1000)
end)

test("performance: many lookups", function()
  local h = caml_hash_create()
  for i = 1, 100 do
    caml_hash_add(h, i, i * 2)
  end
  for i = 1, 100 do
    local value = caml_hash_find(h, i)
    assert_eq(value, i * 2)
  end
end)

test("performance: mixed operations", function()
  local h = caml_hash_create()
  for i = 1, 100 do
    caml_hash_add(h, i, i)
  end
  for i = 1, 50 do
    caml_hash_remove(h, i)
  end
  for i = 51, 100 do
    assert_true(caml_hash_mem(h, i))
  end
  assert_eq(caml_hash_length(h), 50)
end)

print()
print(string.rep("=", 60))
print("Tests passed: " .. tests_passed)
print("Tests failed: " .. tests_failed)
if tests_failed == 0 then
  print("All tests passed! ✓")
  print(string.rep("=", 60))
  os.exit(0)
else
  print("Some tests failed.")
  print(string.rep("=", 60))
  os.exit(1)
end

File: compiler/tests-lua/stdlib_audit/test_hashtbl.ml
(* Comprehensive Hashtbl module test for lua_of_ocaml *)

let () =
  print_endline "=== Comprehensive Hashtbl Module Test Suite ===";
  print_endline ""

(* ==== Hashtbl Creation ==== *)
let () =
  print_endline "--- Hashtbl Creation ---";
  let h = Hashtbl.create 10 in
  Printf.printf "Hashtbl.create 10: length %d\n" (Hashtbl.length h);

  let h2 = Hashtbl.create 0 in
  Printf.printf "Hashtbl.create 0: length %d (should default to valid size)\n" (Hashtbl.length h2);
  print_endline ""

(* ==== Add and Find ==== *)
let () =
  print_endline "--- Add and Find ---";
  let h = Hashtbl.create 10 in
  Hashtbl.add h "key1" "value1";
  Hashtbl.add h "key2" "value2";
  Hashtbl.add h "key3" "value3";
  Printf.printf "After adding 3 keys: length %d\n" (Hashtbl.length h);

  Printf.printf "find \"key1\": %s\n" (Hashtbl.find h "key1");
  Printf.printf "find \"key2\": %s\n" (Hashtbl.find h "key2");
  Printf.printf "find \"key3\": %s\n" (Hashtbl.find h "key3");
  print_endline ""

(* ==== Integer Keys ==== *)
let () =
  print_endline "--- Integer Keys ---";
  let h = Hashtbl.create 10 in
  Hashtbl.add h 1 "one";
  Hashtbl.add h 2 "two";
  Hashtbl.add h 3 "three";
  Printf.printf "find 1: %s\n" (Hashtbl.find h 1);
  Printf.printf "find 2: %s\n" (Hashtbl.find h 2);
  Printf.printf "find 3: %s\n" (Hashtbl.find h 3);
  print_endline ""

(* ==== Find_opt ==== *)
let () =
  print_endline "--- Find_opt ---";
  let h = Hashtbl.create 10 in
  Hashtbl.add h "exists" "value";
  (match Hashtbl.find_opt h "exists" with
   | Some v -> Printf.printf "find_opt \"exists\": Some %s\n" v
   | None -> print_endline "find_opt \"exists\": None");
  (match Hashtbl.find_opt h "missing" with
   | Some v -> Printf.printf "find_opt \"missing\": Some %s\n" v
   | None -> print_endline "find_opt \"missing\": None");
  print_endline ""

(* ==== Mem ==== *)
let () =
  print_endline "--- Mem ---";
  let h = Hashtbl.create 10 in
  Hashtbl.add h "key1" "value1";
  Printf.printf "mem \"key1\": %b\n" (Hashtbl.mem h "key1");
  Printf.printf "mem \"missing\": %b\n" (Hashtbl.mem h "missing");
  print_endline ""

(* ==== Remove ==== *)
let () =
  print_endline "--- Remove ---";
  let h = Hashtbl.create 10 in
  Hashtbl.add h "key1" "value1";
  Hashtbl.add h "key2" "value2";
  Printf.printf "Before remove: length %d\n" (Hashtbl.length h);
  Hashtbl.remove h "key1";
  Printf.printf "After removing key1: length %d\n" (Hashtbl.length h);
  Printf.printf "mem \"key1\" after remove: %b\n" (Hashtbl.mem h "key1");
  Printf.printf "mem \"key2\" still there: %b\n" (Hashtbl.mem h "key2");
  print_endline ""

(* ==== Replace ==== *)
let () =
  print_endline "--- Replace ---";
  let h = Hashtbl.create 10 in
  Hashtbl.add h "key1" "old_value";
  Printf.printf "Before replace: %s\n" (Hashtbl.find h "key1");
  Hashtbl.replace h "key1" "new_value";
  Printf.printf "After replace: %s\n" (Hashtbl.find h "key1");
  Printf.printf "Length after replace: %d\n" (Hashtbl.length h);

  (* Replace on new key should add it *)
  Hashtbl.replace h "key2" "value2";
  Printf.printf "After replacing new key: length %d\n" (Hashtbl.length h);
  print_endline ""

(* ==== Duplicate Keys ==== *)
let () =
  print_endline "--- Duplicate Keys (add multiple times) ---";
  let h = Hashtbl.create 10 in
  Hashtbl.add h "dup" "first";
  Hashtbl.add h "dup" "second";
  Hashtbl.add h "dup" "third";
  Printf.printf "After adding 3 times: length %d\n" (Hashtbl.length h);
  Printf.printf "find returns most recent: %s\n" (Hashtbl.find h "dup");
  print_endline ""

(* ==== Find_all ==== *)
let () =
  print_endline "--- Find_all ---";
  let h = Hashtbl.create 10 in
  Hashtbl.add h "multi" "first";
  Hashtbl.add h "multi" "second";
  Hashtbl.add h "multi" "third";
  let values = Hashtbl.find_all h "multi" in
  Printf.printf "find_all \"multi\" returns %d values: " (List.length values);
  List.iter (fun v -> Printf.printf "%s; " v) values;
  print_endline "";
  print_endline ""

(* ==== Clear ==== *)
let () =
  print_endline "--- Clear ---";
  let h = Hashtbl.create 10 in
  Hashtbl.add h "k1" "v1";
  Hashtbl.add h "k2" "v2";
  Hashtbl.add h "k3" "v3";
  Printf.printf "Before clear: length %d\n" (Hashtbl.length h);
  Hashtbl.clear h;
  Printf.printf "After clear: length %d\n" (Hashtbl.length h);
  Printf.printf "mem \"k1\" after clear: %b\n" (Hashtbl.mem h "k1");
  print_endline ""

(* ==== Iter ==== *)
let () =
  print_endline "--- Iter ---";
  let h = Hashtbl.create 10 in
  Hashtbl.add h "a" 1;
  Hashtbl.add h "b" 2;
  Hashtbl.add h "c" 3;
  print_endline "Iterating over entries:";
  Hashtbl.iter (fun k v -> Printf.printf "  %s -> %d\n" k v) h;
  print_endline ""

(* ==== Fold ==== *)
let () =
  print_endline "--- Fold ---";
  let h = Hashtbl.create 10 in
  Hashtbl.add h "a" 10;
  Hashtbl.add h "b" 20;
  Hashtbl.add h "c" 30;
  let sum = Hashtbl.fold (fun _k v acc -> acc + v) h 0 in
  Printf.printf "Sum of values via fold: %d\n" sum;

  let count = Hashtbl.fold (fun _k _v acc -> acc + 1) h 0 in
  Printf.printf "Count via fold: %d\n" count;
  print_endline ""

(* ==== Copy ==== *)
let () =
  print_endline "--- Copy ---";
  let h1 = Hashtbl.create 10 in
  Hashtbl.add h1 "key1" "value1";
  Hashtbl.add h1 "key2" "value2";
  let h2 = Hashtbl.copy h1 in
  Printf.printf "Original length: %d\n" (Hashtbl.length h1);
  Printf.printf "Copy length: %d\n" (Hashtbl.length h2);

  (* Modify copy, should not affect original *)
  Hashtbl.add h2 "key3" "value3";
  Printf.printf "After adding to copy: original %d, copy %d\n"
    (Hashtbl.length h1) (Hashtbl.length h2);
  print_endline ""

(* ==== Length ==== *)
let () =
  print_endline "--- Length ---";
  let h = Hashtbl.create 10 in
  Printf.printf "Empty: length %d\n" (Hashtbl.length h);
  Hashtbl.add h "k1" "v1";
  Printf.printf "After 1 add: length %d\n" (Hashtbl.length h);
  Hashtbl.add h "k2" "v2";
  Hashtbl.add h "k3" "v3";
  Printf.printf "After 3 adds: length %d\n" (Hashtbl.length h);
  Hashtbl.remove h "k2";
  Printf.printf "After 1 remove: length %d\n" (Hashtbl.length h);
  print_endline ""

(* ==== Large Hashtbl (resize test) ==== *)
let () =
  print_endline "--- Large Hashtbl (automatic resize) ---";
  let h = Hashtbl.create 4 in
  for i = 1 to 100 do
    Hashtbl.add h i (i * 10)
  done;
  Printf.printf "After adding 100 entries: length %d\n" (Hashtbl.length h);

  (* Verify all entries are findable *)
  let all_found = ref true in
  for i = 1 to 100 do
    if Hashtbl.find h i <> i * 10 then all_found := false
  done;
  Printf.printf "All 100 entries findable: %b\n" !all_found;
  print_endline ""

(* ==== Mixed Types ==== *)
let () =
  print_endline "--- Mixed Types (polymorphic) ---";
  let h_int = Hashtbl.create 10 in
  Hashtbl.add h_int 42 "answer";
  Hashtbl.add h_int 7 "lucky";
  Printf.printf "Integer key 42: %s\n" (Hashtbl.find h_int 42);

  let h_float = Hashtbl.create 10 in
  Hashtbl.add h_float 3.14 "pi";
  Hashtbl.add h_float 2.71 "e";
  Printf.printf "Float key 3.14: %s\n" (Hashtbl.find h_float 3.14);
  print_endline ""

(* ==== Complex Values ==== *)
let () =
  print_endline "--- Complex Values ---";
  let h1 = Hashtbl.create 10 in
  Hashtbl.add h1 "list" [1; 2; 3; 4; 5];
  let list_val = Hashtbl.find h1 "list" in
  Printf.printf "List value length: %d\n" (List.length list_val);

  let h2 = Hashtbl.create 10 in
  Hashtbl.add h2 "tuple" [(1, "one"); (2, "two")];
  let tuple_val = Hashtbl.find h2 "tuple" in
  Printf.printf "Tuple list length: %d\n" (List.length tuple_val);
  print_endline ""

(* ==== Not_found Exception ==== *)
let () =
  print_endline "--- Not_found Exception ---";
  let h = Hashtbl.create 10 in
  Hashtbl.add h "exists" "value";
  (try
    let _ = Hashtbl.find h "missing" in
    print_endline "ERROR: Should have raised Not_found"
  with Not_found ->
    print_endline "Correctly raised Not_found for missing key");
  print_endline ""

(* ==== Reset ==== *)
let () =
  print_endline "--- Reset ---";
  let h = Hashtbl.create 4 in
  (* Add many elements to cause resize *)
  for i = 1 to 20 do
    Hashtbl.add h i (i * 10)
  done;
  Printf.printf "After 20 adds: length %d\n" (Hashtbl.length h);
  Hashtbl.reset h;
  Printf.printf "After reset: length %d\n" (Hashtbl.length h);
  (* Can still use after reset *)
  Hashtbl.add h 100 999;
  Printf.printf "After adding to reset table: length %d\n" (Hashtbl.length h);
  print_endline ""

(* ==== Filter_map_inplace ==== *)
let () =
  print_endline "--- Filter_map_inplace ---";
  let h = Hashtbl.create 10 in
  Hashtbl.add h "a" 1;
  Hashtbl.add h "b" 2;
  Hashtbl.add h "c" 3;
  Hashtbl.add h "d" 4;
  Printf.printf "Before filter_map_inplace: length %d\n" (Hashtbl.length h);
  (* Keep even values, double them *)
  Hashtbl.filter_map_inplace (fun _k v ->
    if v mod 2 = 0 then Some (v * 2) else None
  ) h;
  Printf.printf "After filter_map_inplace: length %d\n" (Hashtbl.length h);
  print_endline "Remaining entries:";
  Hashtbl.iter (fun k v -> Printf.printf "  %s -> %d\n" k v) h;
  print_endline ""

(* ==== Stats ==== *)
let () =
  print_endline "--- Stats ---";
  let h = Hashtbl.create 10 in
  for i = 1 to 50 do
    Hashtbl.add h i (i * 2)
  done;
  let stats = Hashtbl.stats h in
  Printf.printf "Stats for 50-entry hashtbl:\n";
  Printf.printf "  num_bindings: %d\n" stats.num_bindings;
  Printf.printf "  num_buckets: %d\n" stats.num_buckets;
  Printf.printf "  max_bucket_length: %d\n" stats.max_bucket_length;
  Printf.printf "  bucket_histogram length: %d\n" (Array.length stats.bucket_histogram);
  print_endline ""

(* ==== Edge Cases ==== *)
let () =
  print_endline "--- Edge Cases ---";

  (* Empty string key *)
  let h = Hashtbl.create 10 in
  Hashtbl.add h "" "empty_key";
  Printf.printf "Empty string key: %s\n" (Hashtbl.find h "");

  (* Zero as key *)
  let h2 = Hashtbl.create 10 in
  Hashtbl.add h2 0 "zero";
  Printf.printf "Zero key: %s\n" (Hashtbl.find h2 0);

  (* Remove non-existent (should not error) *)
  Hashtbl.remove h "non_existent";
  Printf.printf "Remove non-existent: no error\n";

  print_endline ""

(* ==== Stress Test ==== *)
let () =
  print_endline "--- Stress Test ---";
  let h = Hashtbl.create 10 in

  (* Add 1000 entries *)
  for i = 1 to 1000 do
    Hashtbl.add h i (i * 3)
  done;

  (* Verify all can be found *)
  let errors = ref 0 in
  for i = 1 to 1000 do
    try
      let v = Hashtbl.find h i in
      if v <> i * 3 then incr errors
    with Not_found -> incr errors
  done;
  Printf.printf "Added 1000 entries, lookup errors: %d\n" !errors;

  (* Remove every other entry *)
  for i = 1 to 1000 do
    if i mod 2 = 0 then Hashtbl.remove h i
  done;
  Printf.printf "After removing evens: length %d\n" (Hashtbl.length h);

  (* Verify removed entries are gone *)
  let found_removed = ref 0 in
  for i = 2 to 1000 do
    if i mod 2 = 0 && Hashtbl.mem h i then incr found_removed
  done;
  Printf.printf "Should have removed evens, but found: %d\n" !found_removed;

  print_endline ""

(* ==== Final Summary ==== *)
let () =
  print_endline "=== Test Suite Complete ===";
  print_endline "All Hashtbl operations tested successfully!"
